### PR TITLE
feat: Add OpenAPI

### PR DESCRIPTION
**Pull request by provokateurin**
Closes https://github.com/nextcloud/groupfolders/issues/3208
~This is a "breaking" change to the API, but since it was not usable before due to CSRF restrictions it doesn't matter.
While working on this I also had to cleanup a lot of the internal typings which resulted in this relatively large change for adding OpenAPI.
If this is too big please let me know and I will try to split it up.~

### CODE REVIEW

File: README.md
Comment (come-nc) [inline README.md:129]: Hum any chance at linking to a readable form of the doc?

File: lib/Command/ListCommand.php
Comment (come-nc) [inline lib/Command/ListCommand.php:90]: Why do you remove group_details? This is a breaking change for any script relying on that.

File: lib/Command/ListCommand.php
Comment (come-nc) [inline lib/Command/ListCommand.php:113]:
```suggestion
					$folder['groups'] = '';
```

File: lib/Folder/Folder.php
Comment (come-nc) [inline lib/Folder/Folder.php:14]: Avoid using the same name as core classes, it’s confusing when the class is imported with a `use` statement.

File: lib/Folder/FolderManager.php
Comment (come-nc) [inline lib/Folder/FolderManager.php:74]: This is a controller specific thing that ends up in the manager, is there no way to avoid that?

File: lib/Folder/FolderMapping.php
Comment (come-nc) [inline lib/Folder/FolderMapping.php:12]: These classes should actually be entities with mappers, no?

File: lib/Command/ListCommand.php
Comment (provokateurin) [reply lib/Command/ListCommand.php:90]: Yes, as I said at the top.
This was discussed in last weeks team call with @sorbaugh.
It is fine to do this once for the benefit of a much cleaner API.

File: README.md
Comment (provokateurin) [reply README.md:129]: Not sure how that would work since we do not have an existing documentation page for Groupfolders.

Comment (artonge) [reply README.md:129]: Can you generate a `.md` files from the openapi doc? Then we could link to that?

File: lib/Folder/FolderMapping.php
Comment (provokateurin) [reply lib/Folder/FolderMapping.php:12]: I guess so. I only made this change from array to class last minute, so I didn't give it much thought.

File: lib/Folder/Folder.php
Comment (provokateurin) [reply lib/Folder/Folder.php:14]: But `Groupfolder` is also not a good name. Would `InternalFolder` be better as it is only used as an internal type between different methods?

**Comment by provokateurin** [comment]
I will try to extract some of the changes into separate PRs to get them merged first and then come back to this PR.

File: lib/Folder/Folder.php
Comment (come-nc) [reply lib/Folder/Folder.php:14]: As said below this should be in Db and be an entity with a mapper ideally.
I do think Groupfolder is closer to what it is than Folder.
Could be FolderDefinition or FolderSettings if that makes sense?

File: README.md
Comment (come-nc) [reply README.md:129]: I was thinking of a magic link like https://petstore.swagger.io/?url=https://raw.githubusercontent.com/nextcloud/groupfolders/ac3b696dbb820a5ae34039be1300c1309e58d4e1/openapi.json
tables application does that.
Using this instance is not ideal as it’s supposed to be a demo and it may even be contrary to their ToS, I did not check. But maybe there is a similar tool we can use, or we can host our own?

File: lib/Folder/FolderMapping.php
Comment (provokateurin) [reply lib/Folder/FolderMapping.php:12]: Moved to https://github.com/nextcloud/groupfolders/issues/3389, I will revert these changes in this PR

**Comment by provokateurin** [comment]
I removed everything not strictly necessary to implement this. Some improvements can just be done later, like https://github.com/nextcloud/openapi-extractor/issues/172.

File: lib/Folder/FolderManager.php
Comment (come-nc) [inline lib/Folder/FolderManager.php:517]: A bit too much? :open_mouth:

**Review by come-nc — COMMENTED** [review]
Why removing all the id indexing, is that so bad?
Also, why the snake_case on string ids, do we have a convention about that in OCS?

**Comment by provokateurin** [comment]
> Why removing all the id indexing, is that so bad?

Not really, but I just removed it because it is never used.
I prefer a `list` strongly over `array` if possible, due to the ambiguities that can happen.
I can also revert it, if you really don't like it.

>  Also, why the snake_case on string ids, do we have a convention about that in OCS?

No, I just wanted to make them all consistent.
I personal prefer snake_case for JSON, but we can also go with camelCase.

**Comment by susnux** [comment]
> I personal prefer snake_case for JSON, but we can also go with camelCase.

I think we use camelCase for both PHP and JS variables and for JS working with camelCase JSON is more convenient. So if we somewhen might consider a codestyle also JSON camelCase would be nice

**Comment by provokateurin** [comment]
> I think we use camelCase for both PHP and JS variables and for JS working with camelCase JSON is more convenient. So if we somewhen might consider a codestyle also JSON camelCase would be nice

Ok, fair enough :)

**Comment by provokateurin** [comment]
> Why removing all the id indexing, is that so bad?

Also the code on the frontend side was very hacky in this regard, it is a lot better now.

**Comment by provokateurin** [comment]
Depends on https://github.com/nextcloud/groupfolders/pull/3407 and https://github.com/nextcloud/openapi-extractor/pull/181.
No breaking changes anymore, but also very ugly typing :see_no_evil:

**Comment by provokateurin** [comment]
All depending PRs are merged, so this is finally ready.

**Review by sorbaugh — APPROVED** [review]

**Comment by provokateurin** [comment]
Still waiting for a review from @come-nc before merging to ensure I addressed all previous problems.

**Comment by provokateurin** [comment]
Thanks a lot everyone! :tada:

**Pull request merged**
